src/hypothesis/extra/datetime.py

```diff
@@ -23,62 +32,185 @@
+    msg = ('The %s argument is deprecated.  Please use %s_datetime '
+           'with a datetime.datetime or datetime.date object instead.'
+           ) % (name, name[:3])
+    if no_new_args:
+        note_deprecation(msg)
+    else:
+        msg += '  It is an error to use new and deprecated arguments.'
+        raise InvalidArgument(msg)
+    check_valid_interval(dt.MINYEAR, year, 'dt.MINYEAR', name)
+    check_valid_interval(year, dt.MAXYEAR, name, 'dt.MAXYEAR')
+    to = dt.datetime.max if name == 'max_year' else dt.datetime.min
+    return to.replace(year=year)
+
+
+def ensure_datetime_bound(val, name):
+    """Ensure that the given value is a datetime, naive or in UTC."""
+    assert name in ('min_datetime', 'max_datetime')
+    default = dt.datetime.max if 'max_' in name else dt.datetime.min
+    if val is None:
+        return default
+    if not isinstance(val, dt.date):
+        raise InvalidArgument(
+            '%s=%r is not a date or datetime object' % (name, val))
+    if not isinstance(val, dt.datetime):
+        return default.replace(val.year, val.month, val.day)
+    if val.tzinfo is not None:
+        try:
+            return val.astimezone(pytz.utc)
+        except OverflowError:
+            raise InvalidArgument(
+                '%s=%r overflows when converted to UTC' % (name, val))
+    return val
+
 
 class DatetimeStrategy(SearchStrategy):
 
-    def __init__(self, allow_naive, timezones, min_year=None, max_year=None):
+    def __init__(self, allow_naive, timezones, min_year=None, max_year=None,
+                 min_datetime=None, max_datetime=None):
+        if not (allow_naive or timezones):
+            raise InvalidArgument('Cannot create non-naive datetimes with '
+                                  'no timezones allowed')
         self.allow_naive = allow_naive
         self.timezones = timezones
-        self.min_year = min_year or dt.MINYEAR
-        self.max_year = max_year or dt.MAXYEAR
-        for a in ['min_year', 'max_year']:
-            year = getattr(self, a)
-            if year < dt.MINYEAR:
-                raise InvalidArgument(u'%s out of range: %d < %d' % (
-                    a, year, dt.MINYEAR
-                ))
-            if year > dt.MAXYEAR:
-                raise InvalidArgument(u'%s out of range: %d > %d' % (
-                    a, year, dt.MAXYEAR
-                ))
+
+        old = min_datetime is None and max_datetime is None
+        if min_year is not None:
+            min_datetime = convert_year_to_datetime(min_year, 'min_year', old)
+        if max_year is not None:
+            max_datetime = convert_year_to_datetime(max_year, 'max_year', old)
+        self.min_dt = ensure_datetime_bound(min_datetime, 'min_datetime')
+        self.max_dt = ensure_datetime_bound(max_datetime, 'max_datetime')
+
+        if (self.min_dt.tzinfo is None) != (self.max_dt.tzinfo is None):
+            # Fix the "only supplied one bound" case, or raise InvalidArgument
```

Comment:
Done.  Also fixed the check, which should be `*_datetime is None` to avoid mishandling mixed bounds where the naive one is an extreme.